Approved/multidimensional-arrays.md

```diff
@@ -0,0 +1,657 @@
+
+### Examples
+
+Example 1:
+Representing a complex-valued matrix with jagged arrays.
+
+```qsharp
+open Microsoft.Quantum.Math;
+
+let y = [
+    [Complex(0.0, 0.0), Complex(0.0, -1.0)],
+    [Complex(0.0, 1.0), Complex(0.0, 0.0)]
+];
+// y: Complex[][]
+let upperRight = y[0][1];
+// upperRight: Complex = Complex(0.0, -1.0);
+```
+
+Example 2:
+Representing the generators of a stabilizer group as dense arrays of Pauli operators.
+
+```qsharp
+open Microsoft.Quantum.Arrays;
+
+let perfect = [
+    [PauliI, PauliX, PauliZ, PauliZ, PauliX],
+    [PauliX, PauliZ, PauliZ, PauliX, PauliI],
+    [PauliZ, PauliZ, PauliX, PauliI, PauliX],
+    [PauliZ, PauliX, PauliI, PauliX, PauliZ]
+];
+// perfect: Pauli[][]
+let syndrome = ForEach(Measure(_, register), perfect);
+// syndrome: Result[]
+```
+
+Example 3:
+Representing the generators of a stabilizer group as a sparse array of Pauli operators and their indices.
+
+```qsharp
+let shor = [
+    [(PauliX, 0), (PauliX, 1)],
+    [(PauliX, 1), (PauliX, 2)],
+    [(PauliX, 3), (PauliX, 4)],
+    [(PauliX, 4), (PauliX, 5)],
+    [(PauliX, 6), (PauliX, 7)],
+    [(PauliX, 7), (PauliX, 8)],
+
+    [(PauliZ, 0), (PauliZ, 1), (PauliZ, 2), (PauliZ, 3), (PauliZ, 4), (PauliZ, 6)],
+    [(PauliZ, 3), (PauliZ, 4), (PauliZ, 6), (PauliZ, 7), (PauliZ, 8), (PauliZ, 9)],
+];
+// shor: (Pauli, Int)[][]
+```
+
+Example 4:
+A prior distribution over phases represented as an array of double-precision real numbers.
+
+```qsharp
+open Microsoft.Quantum.Arrays;
+open Microsoft.Quantum.Convert;
+
+// suppose likelihood is a function of type Int -> Double.
+let nPoints = 1000;
+let prior = ConstantArray(nPoints, 1.0 / IntAsDouble(nPoints));
+mutable posterior = prior;
+for ((idx, pr) in Enumerated(posterior)) {
+    set posterior w/= idx <- posterior[idx] * likelihood(idx);
+}
+set posterior = PNormalized(posterior, 1.0);
+```
+
+Example 5:
+A function that returns an elementary matrix (that is, a matrix with a one in the (𝑖, 𝑗)th element and zeros elsewhere).
+
+```qsharp
+open Microsoft.Quantum.Arrays;
+
+function ElementaryMatrix<'T>(
+    (nRows : Int, nColumns : Int),
+    (idxRow : Int, idxCol : Int),
+    zero : 'T, one : 'T
+) : 'T[][] {
+    mutable matrix = ConstantArray(
+        nRows,
+        ConstantArray(nCols, zero)
+    );
+    return matrix w/ row <- (matrix[row] w/ col <- one);
+}
+
+let e01 = ElementaryMatrix((2, 2), (0, 1), 0.0, 1.0);
+// e01: Double[][] = [[0.0, 1.0], [0.0, 0.0]]
+```
+
+## Proposed Modification
+
+Building on the utility of 1-D array notation, this suggestion proposes modifying Q# to include new multidimensional array types `'T[,]`, `'T[,,]`, and so forth. Like values of type `'T[]`, these new multidimensional would also be immutable, and could be manipulated by using the subscript (`[]`) and copy-and-update (`w/`) operators.
```

Comment:
I think we have an opportunity to simplify our array type syntax for both one- and multi-dimensional arrays. It would be nice to have just one syntax for both types and values. For example, instead of

* `[1, 2, 3]` is of type `Int[]`
* `[[1, 2, 3]]` is of type `Int[][]`
* `#[[1, 2, 3], [4, 5, 6]]` is of type `Int[,]`
* `##[[[1, 2, 3]]] is of type `Int[,,]`

why not instead:

* `[1, 2, 3]` is of type `[Int]`
* `[[1, 2, 3]]` is of type `[[Int]]`
* `#[[1, 2, 3], [4, 5, 6]]` is of type `#[Int]`
* `##[[[1, 2, 3]]]` is of type `##[Int]`

or alternatively, if we use a different multidimensional array literal syntax:

* `#[[1, 2, 3], [4, 5, 6]]` is of type `#[Int]`
* `#[#[[1, 2, 3]]]` is of type `#[#[Int]]`

I think this syntax is simpler and more predictable between values and types. It's analogous to tuples, which already follow this pattern. For example, `("foo", 4)` is of type `(String, Int)`. `[T]` also has the advantage of avoiding awkward constructs like `(Int -> Int)[]`, instead allowing more readable and easier-to-parse `[Int -> Int]`.

@bettinaheim @cgranade

Reply:
I like the symmetry of `[Int]` and `(Int)`. It also reduces as you say the cognitive load of `Int[,,]` being the type and `##[ ]` for the literal declaration. Much easier to remember when the type and the way I need to declare a literal are the same/similar :)